src/read.rs

```diff
@@ -1684,18 +1677,221 @@ pub fn read_zipfile_from_stream<'a, R: Read>(reader: &'a mut R) -> ZipResult<Opt
+            }
+
+            match self.read_a_byte()? {
+                None => return Ok(index),
+                Some(value) => buf[index] = value,
+            }
+        }
+        Ok(buf.len())
+    }
+}
+
+impl<T: Read> ReadAndSupplyExpectedCRC32 for ReadTillDataDescriptor<T> {
+    fn get_crc32(&self) -> io::Result<u32> {
+        Ok(self
+            .get_data_descriptor()
+            .ok_or(std::io::Error::new(
+                ErrorKind::Other,
+                "EOF reached, corrupt zip file",
+            ))?
+            .crc32)
+    }
+}
+
+/// Read ZipFile structures from a reader.
+///
+/// This method operates in three different modes:
+/// 1. data_descriptor=None and allow_unlimited_read=false : Parse the local file header and return zip file. Works if data descriptors are not used.
+/// 2. data_descriptor=Some and allow_unlimited_read=false : Parse the local file header, augment the zip file with the size given in the data descriptor and return zip file. Works if data descriptors are used.
+/// 3. data_descriptor=None and allow_unlimited_read=true : Parse the local file header, if data descriptors not used, fallback to case 1, else provide a ZipFile that
+///     searches for the data descriptor in the file. It just starts reading the compressed stream with no size limit set. When detecting a data descriptor the
+///     stream is closed. This is useful when the size of the compressed stream is unknown but brings some security risks. Therefore, returning an UntrustedValue.
+/// 4. data_descriptor=Some and allow_unlimited_read=true : Fallback to case 2
+///
+/// Case 3: **Security-disclaimer**: The **output** should **not** be regarded as **secure/trustworthy**.
+/// When an attacker has control over a file which is compressed into a zip: The file might be crafted such that
+/// when reading with ZipArchive vs this function a parsing mismatch can occur. An attacker that has control over a file
+/// can manipulate such file such that this function will regard parts of the file as new zip file entry. The attacker might
+/// therefore inject custom zip file entries with attacker controlled content/name/metadata.
+fn read_zipfile_from_fileblock<'a, R: Read>(
+    block: ZipLocalEntryBlockAndFields,
+    reader: &'a mut R,
+    data_descriptor: Option<ZipDataDescriptor>,
+    mut allow_unlimited_read: bool,
+) -> ZipResult<MaybeUntrusted<Option<ZipFile<'_>>>> {
+    if data_descriptor.is_some() && allow_unlimited_read {
+        allow_unlimited_read = false; // fallback, size is given
+    }
+
+    let size_unknown = data_descriptor.is_none() && block.block.flags & (1 << 3) == 1 << 3; // no data descriptor provided but should exist
+    let mut result = ZipFileData::from_local_block(block, data_descriptor, allow_unlimited_read)?;
+
+    if size_unknown && !allow_unlimited_read {
+        return Err(ZipError::UnsupportedArchive(
+            "Archive is using data descriptors, but unlimited reading not enabled",
+        ));
+    }
```

Comment:
We're not using `result` in this case, so we can handle it slightly faster and be more readable by moving it above the `from_local_block` call.
```suggestion
    if size_unknown && !allow_unlimited_read {
        return Err(ZipError::UnsupportedArchive(
            "Archive is using data descriptors, but unlimited reading not enabled",
        ));
    }

    let mut result = ZipFileData::from_local_block(block, data_descriptor, allow_unlimited_read)?;
```